crates/services/upgradable-executor/src/executor.rs

```diff
@@ -909,39 +976,76 @@ where
         mode: ProduceBlockMode,
         new_tx_waiter: impl NewTxWaiterPort,
         preconfirmation_sender: impl PreconfirmationSenderPort,
-    ) -> ExecutorResult<Uncommitted<ExecutionResult, Changes>>
+    ) -> ExecutorResult<
+        Uncommitted<(ExecutionResult, Vec<StorageReadReplayEvent>), Changes>,
+    >
     where
         TxSource: TransactionsSource + Send + Sync + 'static,
     {
         let relayer = self.relayer_view_provider.latest_view()?;
 
         let db_height = match mode {
             ProduceBlockMode::Produce => block.header_to_produce.height().pred(),
-            ProduceBlockMode::DryRunLatest => None,
-            ProduceBlockMode::DryRunAt { height } => height.pred(),
+            ProduceBlockMode::DryRun { height, .. } => match height {
+                BlockHeightSelection::Latest => None,
+                BlockHeightSelection::Past { height } => height.pred(),
+            },
         };
 
-        if let Some(previous_block_height) = db_height {
+        let mut storage_rec = Default::default();
+
+        let result = if let Some(previous_block_height) = db_height {
```

Comment:
```suggestion
        let height = db_height
                .as_ref()
                .map_or_else(
                        || self.storage_view_provider.latest_view(),
                        |h| self.storage_view_provider.view_at(h),
                )?;

        let database = if mode.record_storage_reads() {
                let recorder = StorageAccessRecorder::new(height);
                storage_rec = recorder.record.clone();
                recorder.into()
        } else {
                height
        };

        let result = ExecutionInstance::new(relayer, database, options)
                .produce_without_commit(
                        block,
                        mode.is_dry_run(),
                        new_tx_waiter,
                        preconfirmation_sender,
                )
                .await;
```

would this work instead?

Reply:
No. Database with a recorder has a separate type, and cannot be held in the same variable as a db without recorder.